File: python/navigation/local_planning/predictive/mpc.py
"""Model Predictive Control for a kinematic mobile robot (Klančar & Škrjanc,
Robotics and Autonomous Systems 55(6):460-469, 2007, DOI
10.1016/j.robot.2007.01.002): each tick predicts the next H steps with the
unicycle model, optimizes the control sequence U against the shared receding-
horizon cost J(U), executes only the first control u_0, and re-optimizes next
tick (receding horizon).

The optimizer is fixed-iteration projected gradient descent with a central
finite-difference gradient. Finite differences (rather than a hand-derived
analytic Jacobian of the arc integrator) are chosen because the gradient then
follows mechanically from the one shared scalar cost -- so as long as J(U) is
identical across Python/C++/TS, the descent is too, and the contrast with MPPI
stays a clean "same J, sampling instead of gradient".
"""

from __future__ import annotations

from navigation.core.capabilities import Capability, ObstacleQuery
from navigation.core.params import ParamSet
from navigation.core.planner import ObstacleLocalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import LocalTask, Pose, RobotState, VelocityCommand

from ._rollout import clamp, rollout, sequence_cost

# A control sequence U = [u_0, ..., u_{H-1}] with u_k = (v_k, omega_k).
_Controls = list[tuple[float, float]]


class MpcPlanner(ObstacleLocalPlanner):
    def __init__(self, params: ParamSet) -> None:
        super().__init__(params)
        self._horizon = params.get_int("horizon")
        self._iterations = params.get_int("iterations")
        self._step_alpha = params.get_float("step_alpha")
        self._grad_eps = params.get_float("grad_eps")
        self._max_step_v = params.get_float("max_step_v")
        self._max_step_omega = params.get_float("max_step_omega")
        self._w_goal = params.get_float("w_goal")
        self._w_obstacle = params.get_float("w_obstacle")
        self._w_control = params.get_float("w_control")
        self._min_obstacle_dist = params.get_float("min_obstacle_dist")
        self._v_max = params.get_float("v_max")
        self._omega_max = params.get_float("omega_max")
        self._a_max = params.get_float("a_max")
        self._footprint_radius = params.get_float("footprint_radius")

        # Control sequence U carried across ticks for warm-starting; empty = cold
        # start (first tick / after reset()), which seeds U with zeros.
        self._controls: _Controls = []

    @property
    def name(self) -> str:
        return "mpc"

    def required_capabilities(self) -> set[Capability]:
        return {Capability.OBSTACLE_QUERY}

    def reset(self) -> None:
        self._controls = []

    def _warm_start(self) -> _Controls:
        """Left-shift the previous U and duplicate its last control, so this
        tick starts from where the last horizon left off (the executed u_0 is
        dropped). Cold start seeds H zero controls."""
        if not self._controls:
            return [(0.0, 0.0) for _ in range(self._horizon)]
        return self._controls[1:] + [self._controls[-1]]

    def _cost(
        self, space: ObstacleQuery, s0: Pose, controls: _Controls, goal: Pose, h: float
    ) -> float:
        traj = rollout(s0, controls, h)
        return sequence_cost(
            space,
            traj,
            controls,
            goal,
            self._footprint_radius,
            self._w_goal,
            self._w_obstacle,
            self._min_obstacle_dist,
            self._w_control,
        )

    def _descent_step(
        self, space: ObstacleQuery, s0: Pose, controls: _Controls, goal: Pose, h: float
    ) -> None:
        """One projected gradient-descent iteration: a full 2H-component central
        finite-difference gradient of J at the current U (component order k
        ascending, v before omega -- part of the cross-language determinism
        contract), then one U <- U - step_alpha*grad update with per-component
        step clamp and box projection (v in [0, v_max], |omega| <= omega_max)."""
        n = len(controls)
        eps = self._grad_eps
        grad_v = [0.0] * n
        grad_omega = [0.0] * n
        for k in range(n):
            v, omega = controls[k]
            controls[k] = (v + eps, omega)
            j_plus = self._cost(space, s0, controls, goal, h)
            controls[k] = (v - eps, omega)
            j_minus = self._cost(space, s0, controls, goal, h)
            grad_v[k] = (j_plus - j_minus) / (2.0 * eps)
            controls[k] = (v, omega + eps)
            j_plus = self._cost(space, s0, controls, goal, h)
            controls[k] = (v, omega - eps)
            j_minus = self._cost(space, s0, controls, goal, h)
            grad_omega[k] = (j_plus - j_minus) / (2.0 * eps)
            controls[k] = (v, omega)

        for k in range(n):
            v, omega = controls[k]
            v -= clamp(self._step_alpha * grad_v[k], -self._max_step_v, self._max_step_v)
            omega -= clamp(
                self._step_alpha * grad_omega[k], -self._max_step_omega, self._max_step_omega
            )
            controls[k] = (
                clamp(v, 0.0, self._v_max),
                clamp(omega, -self._omega_max, self._omega_max),
            )

    def _emit_band(
        self,
        recorder: TraceRecorder,
        s0: Pose,
        traj: list[Pose],
        h: float,
        total_cost: float,
    ) -> None:
        band = [[s0[0], s0[1], s0[2], 0.0]]
        for p in traj:
            band.append([p[0], p[1], p[2], h])
        recorder.band_updated(
            band,
            data={
                "iterations": float(self._iterations),
                "horizon": float(self._horizon),
                "total_cost": total_cost,
            },
        )

    def compute_command(
        self,
        space: ObstacleQuery,
        state: RobotState,
        task: LocalTask,
        dt: float,
        recorder: TraceRecorder | None = None,
    ) -> VelocityCommand:
        s0 = state.pose
        goal = task.goal
        # Prediction step equals the control period (dt): predicting and
        # executing on the same discretization keeps the executed u_0 consistent
        # with the horizon it was optimized in.
        h = dt

        controls = self._warm_start()
        for _ in range(self._iterations):
            self._descent_step(space, s0, controls, goal, h)
        self._controls = controls

        if recorder is not None:
            traj = rollout(s0, controls, h)
            total_cost = sequence_cost(
                space,
                traj,
                controls,
                goal,
                self._footprint_radius,
                self._w_goal,
                self._w_obstacle,
                self._min_obstacle_dist,
                self._w_control,
            )
            self._emit_band(recorder, s0, traj, h, total_cost)

        # Executed command: accel-limit the linear speed against the velocity the
        # simulator reports (RobotState.v, like DWA -- no separate _v_prev state),
        # then box-clamp so the executed command is always within limits.
        v0, omega0 = controls[0]
        v0 = clamp(v0, state.v - self._a_max * h, state.v + self._a_max * h)
        v0 = clamp(v0, 0.0, self._v_max)
        omega0 = clamp(omega0, -self._omega_max, self._omega_max)
        return VelocityCommand(v0, omega0)

File: cpp/src/local_planning/predictive/mpc.cpp
#include "navigation/local_planning/predictive/mpc.hpp"

#include <utility>
#include <vector>

namespace navigation::local_planning {

MpcPlanner::MpcPlanner(core::ParamSet params)
    : core::ObstacleLocalPlanner(std::move(params)),
      horizon_(params_.get_int("horizon")),
      iterations_(params_.get_int("iterations")),
      step_alpha_(params_.get_float("step_alpha")),
      grad_eps_(params_.get_float("grad_eps")),
      max_step_v_(params_.get_float("max_step_v")),
      max_step_omega_(params_.get_float("max_step_omega")),
      w_goal_(params_.get_float("w_goal")),
      w_obstacle_(params_.get_float("w_obstacle")),
      w_control_(params_.get_float("w_control")),
      min_obstacle_dist_(params_.get_float("min_obstacle_dist")),
      v_max_(params_.get_float("v_max")),
      omega_max_(params_.get_float("omega_max")),
      a_max_(params_.get_float("a_max")),
      footprint_radius_(params_.get_float("footprint_radius")) {}

std::vector<Control> MpcPlanner::warm_start() const {
  if (controls_.empty()) return std::vector<Control>(static_cast<size_t>(horizon_), Control{0.0, 0.0});
  std::vector<Control> shifted(controls_.begin() + 1, controls_.end());
  shifted.push_back(controls_.back());
  return shifted;
}

double MpcPlanner::cost(core::ObstacleQuery& space, const core::Pose& s0,
                        const std::vector<Control>& controls, const core::Pose& goal,
                        double h) const {
  std::vector<core::Pose> traj = rollout(s0, controls, h);
  return sequence_cost(space, traj, controls, goal, footprint_radius_, w_goal_, w_obstacle_,
                       min_obstacle_dist_, w_control_);
}

void MpcPlanner::descent_step(core::ObstacleQuery& space, const core::Pose& s0,
                              std::vector<Control>& controls, const core::Pose& goal,
                              double h) const {
  const size_t n = controls.size();
  const double eps = grad_eps_;
  std::vector<double> grad_v(n, 0.0), grad_omega(n, 0.0);
  for (size_t k = 0; k < n; ++k) {
    const double v = controls[k].first, omega = controls[k].second;
    controls[k] = Control{v + eps, omega};
    const double jvp = cost(space, s0, controls, goal, h);
    controls[k] = Control{v - eps, omega};
    const double jvm = cost(space, s0, controls, goal, h);
    grad_v[k] = (jvp - jvm) / (2.0 * eps);
    controls[k] = Control{v, omega + eps};
    const double jwp = cost(space, s0, controls, goal, h);
    controls[k] = Control{v, omega - eps};
    const double jwm = cost(space, s0, controls, goal, h);
    grad_omega[k] = (jwp - jwm) / (2.0 * eps);
    controls[k] = Control{v, omega};
  }

  for (size_t k = 0; k < n; ++k) {
    double v = controls[k].first, omega = controls[k].second;
    v -= clamp(step_alpha_ * grad_v[k], -max_step_v_, max_step_v_);
    omega -= clamp(step_alpha_ * grad_omega[k], -max_step_omega_, max_step_omega_);
    controls[k] = Control{clamp(v, 0.0, v_max_), clamp(omega, -omega_max_, omega_max_)};
  }
}

void MpcPlanner::emit_band(core::TraceRecorder& recorder, const core::Pose& s0,
                           const std::vector<core::Pose>& traj, double h, double total_cost) const {
  std::vector<std::vector<double>> band;
  band.reserve(traj.size() + 1);
  band.push_back({s0.x, s0.y, s0.theta, 0.0});
  for (const core::Pose& p : traj) band.push_back({p.x, p.y, p.theta, h});
  recorder.band_updated(band, {{"iterations", static_cast<double>(iterations_)},
                               {"horizon", static_cast<double>(horizon_)},
                               {"total_cost", total_cost}});
}

core::VelocityCommand MpcPlanner::compute_command(core::ObstacleQuery& space,
                                                  const core::RobotState& state,
                                                  const core::LocalTask& task, double dt,
                                                  core::TraceRecorder* recorder) {
  const core::Pose& s0 = state.pose;
  const core::Pose& goal = task.goal;
  // Prediction step equals the control period (dt): predicting and executing on
  // the same discretization keeps the executed u_0 consistent with the horizon
  // it was optimized in.
  const double h = dt;

  std::vector<Control> controls = warm_start();
  for (int i = 0; i < iterations_; ++i) descent_step(space, s0, controls, goal, h);
  controls_ = controls;

  if (recorder) {
    std::vector<core::Pose> traj = rollout(s0, controls, h);
    const double total_cost = sequence_cost(space, traj, controls, goal, footprint_radius_, w_goal_,
                                            w_obstacle_, min_obstacle_dist_, w_control_);
    emit_band(*recorder, s0, traj, h, total_cost);
  }

  // Executed command: accel-limit the linear speed against the velocity the
  // simulator reports (RobotState.v, like DWA -- no separate v_prev_ member),
  // then box-clamp so the executed command is always within limits.
  double v0 = controls[0].first, omega0 = controls[0].second;
  v0 = clamp(v0, state.v - a_max_ * h, state.v + a_max_ * h);
  v0 = clamp(v0, 0.0, v_max_);
  omega0 = clamp(omega0, -omega_max_, omega_max_);
  return core::VelocityCommand{v0, omega0};
}

}  // namespace navigation::local_planning
